rust/connlib/libs/common/src/session.rs

```diff
@@ -51,38 +51,39 @@ pub struct TunnelAddresses {
 
 // Evaluate doing this not static
 /// Traits that will be used by connlib to callback the client upper layers.
-pub trait Callbacks {
+pub trait Callbacks: Clone + Send + Sync {
     /// Called when there's a change in the resource list.
-    fn on_update_resources(resource_list: ResourceList);
+    fn on_update_resources(&self, resource_list: ResourceList);
```

Comment:
This forces interior-mutablity on all implementors. It might be more ergonomic to either have connlib have a single owner of this or make the trait `&mut self` and `Arc<Mutex>` it inside connlib.

Reply:
We decided that the burden of thread-safety/etc. is on the non-Rust side: https://github.com/firezone/product/blob/main/architecture/madr.md#2023_06_14-client-app-call

On the Rust side, the underlying Apple/Android callback handler is just an opaque pointer to some Swift/Kotlin object.